### PR TITLE
add `directory` input

### LINKED ISSUE #6: Support for additional folders that should be covered
In [Trixi.jl](https://github.com/trixi-framework/Trixi.jl) we store example setups that are used for testing in `examples/`, which lives outside the `src/` folder. To cover these files during coverage, we had to implement the coverage action manually in our CI (inspired by this very helpful action here!). However, I was wondering if it would be possible to add support for additional folders using the `with` syntax, such that one could do something like
```yaml
- uses: julia-actions/julia-processcoverage@v1
  with:
    additional-folders: examples, special_folder
```

The relevant section of our workflow file is here:
https://github.com/trixi-framework/Trixi.jl/blob/b4950a0fd8b8055fb99d846dea22276e3d408b07/.github/workflows/ci.yml#L91-L96

### PR DESCRIPTION
**Pull request by ericphanson**
I'm not sure how to test this so I'd appreciate a careful review!

closes #6 (I think)

### CODE REVIEW

**Comment by DilumAluthge** [comment]
I don't know enough about how this action works to know if this is correct.

**Review by sloede — COMMENTED** [review]
@ericphanson thanks a lot for working on this!

To me it seems like the current implementation will not properly fix #6 for multiple folders. With `LCOV.writefile(...)` in https://github.com/julia-actions/julia-processcoverage/pull/8/files#diff-97a9548c114a5bcb76cdafa38a90e79f35b0f627d6189019e315afcf560d250fL11, the file with coverage info is going to be overwritten, as far as I can tell (see also https://github.com/JuliaCI/CoverageTools.jl/blob/8ef9bd3d8548d53f9d20fc5550cfe207d9e3281a/src/lcov.jl#L24-L30). Thus if invoking this action multiple times, the "last one wins". While it does make the action slightly more useful for a single non-top-level directory to be covered, I think it will fail to cover multiple subdirectories.

What we do in Trixi.jl is to call `process_folder` multiple times and then concatenate the results, see also https://github.com/trixi-framework/Trixi.jl/blob/c79524866d5022184f03d0d61eb30f370efa43f0/.github/workflows/ci.yml#L134. 

IMHO, it would also be helpful to also update the README with how to use this new feature with one or more directories explicitly set.

**Comment by SaschaMann** [comment]
Thanks for the review @sloede! I'm not too familiar with coverage processing either, so this really helps :) 

---

`codecov-action` accepts a list of comma-separated paths to the coverage reports. We could copy that style so that

```yaml
- uses: julia-actions/julia-processcoverage@v1
  with:
    directories: src,examples
```

works and writes it all to a single `lcov.info` file.

---

Another alternative would be to name the lcov file after the (sanitised) input directory and set it as output for the action:

```yaml
- uses: julia-actions/julia-processcoverage@v1
  id: examples-coverage
  with:
    directory: examples

- uses: julia-actions/julia-processcoverage@v1
  id: src-coverage

- uses: codecov/codecov-action@v1
  with:
    files: ${{ steps.examples-coverage.outputs.lcov-files }},${{ steps.src-coverage.outputs.lcov-files }}
```

The second approach would likely also require changes to `julia-actions/julia-uploadcoveralls`

**Comment by sloede** [comment]
@SaschaMann Just from the two code snippets you just showed, I think the first example
```yaml
- uses: julia-actions/julia-processcoverage@v1
  with:
    directories: src,examples
```
is IMHO **much** better in terms of readability and usability. This is also very close to that I had in mind when creating #6.

**Comment by ericphanson** [comment]
Ok, sounds good! So the comma separated paths is not a YAML array, just a string with commas I think? So I can just pass it as an environmental variable still and parse it Julia side. Sounds good, I’ll make that change and update the readme.

**Comment by SaschaMann** [comment]
> So the comma separated paths is not a YAML array, just a string with commas I think?

Yep. Actions only accept strings as input at the moment.

**Comment by sloede** [comment]
> Yep. Actions only accept strings as input at the moment.

One could optionally consider adding the feature that the directory separator can be optionally chosen to allow multiple/non-standard directory names that - for some reason - include a `,`, e.g., `my/really,weird/path`. This could be achieved by adding something like
```yaml
inputs:
  directory:
    description: 'Directory to process. Multiple directories can be specified using a comma-separated list.'
    required: false
    default: 'src'
  separator:
    description: 'Directory separator for specifying multiple directories.'
    required: false
    default: ','
```
But maybe that's overkill, I dunno

**Comment by SaschaMann** [comment]
> But maybe that's overkill, I dunno

Imo if someone has commas in their paths, they can implement the fix for that themselves ¯\_(ツ)_/¯

**Comment by ericphanson** [comment]
Ok, I've made those changes, let me know what you think!

File: README.md [hunk 1]
```diff
@@ -13,3 +13,23 @@ See [PkgTemplates.jl](https://github.com/invenia/PkgTemplates.jl/blob/master/tes
...
 ```
+
+One can also specify the directory or directories to use via the `directories` input (which defaults to `src`). E.g.
+```yaml
+
+      - uses: julia-actions/julia-processcoverage@v1
+      - uses: codecov/codecov-action@v1
+        with:
+          file: lcov.info
+          directories: src,examples
```
Comment (SaschaMann) [inline README.md:24]:
```suggestion
      - uses: julia-actions/julia-processcoverage@v1
        with:
          directories: src,examples
      - uses: codecov/codecov-action@v1
        with:
          file: lcov.info
```

File: README.md [hunk 2]
```diff
@@ -13,3 +13,23 @@ See [PkgTemplates.jl](https://github.com/invenia/PkgTemplates.jl/blob/master/tes
...
+```
+instructs the action to look for coverage information in both `src` and an `examples` folder. Likewise, use
+```yaml
+
+      - uses: julia-actions/julia-processcoverage@v1
+      - uses: codecov/codecov-action@v1
+        with:
+          file: lcov.info
+          directories: path/to/subdir/package/src
```
Comment (SaschaMann) [inline README.md:33]:
```suggestion
      - uses: julia-actions/julia-processcoverage@v1
        with:
          directories: path/to/subdir/package/src
      - uses: codecov/codecov-action@v1
        with:
          file: lcov.info
```

**Review by SaschaMann — COMMENTED** [review]
On first glance it looks good to me.

@sloede could you do a test build with Trixie.jl to see if it works and resolves #6 as you had in mind? If you replace `uses: julia-actions/julia-processcoverage@v1` with `uses: ericphanson/julia-processcoverage@master` on a branch, you should have access to the new input for testing.

File: main.jl [hunk 3]
```diff
@@ -6,6 +6,11 @@ Pkg.add(PackageSpec(name="CoverageTools"))
...
 
-pf = process_folder()
+directories = get(ENV, "INPUT_DIRECTORIES", "src")
+dirs = filter!(!isempty, split(directories, ","))
+for dir in dirs
+    isdir(dir) || error("directory \"$dir\" not found!")
+end
 
-LCOV.writefile("lcov.info", pf)
+pfs = map(process_folder, dirs)
```
Comment (sloede) [inline main.jl:15]: `process_folder` returns a `Vector{FileCoverage}`. Using `map` this way, wouldn't it return a `Vector{Vector{FileCoverage}}`? Shouldn't it be something like
```suggestion
pfs = mapreduce(process_folder, vcat, dirs)
```
instead (although I am not expert on `mapreduce` or `map` :wink:)?

**Review by sloede — CHANGES_REQUESTED** [review]

File: main.jl (hunk 3 above)
Comment (ericphanson) [reply main.jl:15]: ah oops, good point!

**Comment by sloede** [comment]
> could you do a test build with Trixie.jl to see if it works and resolves #6 as you had in mind?

Running in https://github.com/trixi-framework/Trixi.jl/pull/630, but it might take a while...

File: action.yml [hunk 4]
```diff
@@ -6,8 +6,16 @@ branding:
...
+    required: false
+    default: 'src'
+
 runs:
   using: 'composite'
   steps:
     - run: julia --color=yes "$GITHUB_ACTION_PATH"/main.jl
       shell: bash
+      env:
+        - INPUT_DIRECTORIES
```
Comment (sloede) [inline action.yml:21]: This line seems to be wrong: https://github.com/trixi-framework/Trixi.jl/pull/630/checks?check_run_id=2745439242#step:1:34

But I am not an expert  enough on GHA scripting to tell you how to fix it :-/

Comment (ericphanson) [inline action.yml:21]: maybe it should be
```suggestion
      env: INPUT_DIRECTORIES
```
?

Comment (sloede) [reply action.yml:21]: From [here](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#example):
> When you specify an input to an action in a workflow file or use a default input value, GitHub creates an environment variable for the input with the name INPUT_<VARIABLE_NAME>.

Thus maybe this should be just omitted?
```suggestion
```

Comment (ericphanson) [inline action.yml:21]: ah, according to https://github.com/actions/runner/issues/665#issuecomment-676581170, we need to pass it manually like
```suggestion
      env:
        - INPUT_DIRECTORIES: ${{ inputs.directories }}
```

Comment (ericphanson) [reply action.yml]: https://github.com/julia-actions/julia-processcoverage/pull/8/files#r645527013

File: action.yml [hunk 5]
```diff
@@ -6,8 +6,16 @@ branding:
...
+    required: false
+    default: 'src'
+
 runs:
   using: 'composite'
   steps:
     - run: julia --color=yes "$GITHUB_ACTION_PATH"/main.jl
       shell: bash
+      env:
+        - INPUT_DIRECTORIES: ${{ inputs.directories }}
```
Comment (sloede) [inline action.yml:21]:
```suggestion
        INPUT_DIRECTORIES: ${{ inputs.directories }}
```
I believe this is still wrong

Comment (ericphanson) [reply action.yml:21]: thanks!

**Review by sloede — APPROVED** [review]
The results from https://github.com/trixi-framework/Trixi.jl/pull/630 match exactly what we have in `main`, thus I'd say this is ready to be merged.

**Comment by sloede** [comment]
I wonder if it would make sense to tag `v2` after this is merged...

**Comment by SaschaMann** [comment]
> I wonder if it would make sense to tag v2 after this is merged...

Could you elaborate why?

It's only adds to the interface, so I'd usually bump the minor version. I guess technically if `src` doesn't exist it's breaking but I'm not sure if that justifies a major bump? (+ we could change that to a warning, too)

**Comment by sloede** [comment]
Sorry, I just assumed that people only ever go in integer increments over the GH action versions, but of course you are right: This should be a minor version bump.

**Review by SaschaMann — APPROVED** [review]
Thank you all for working on this :) 

I'll merge and tag a release tomorrow morning (so I can revert it in case the `error()` does actually break something across a lot of repos) unless there are any objections until then.

**Comment by ericphanson** [comment]
I know we're already on the same page about minor bump, but I just wanted to add that the error in fact already happens if you don't have a `src` directory:
```julia
julia> process_folder("does_not_exist")
[ Info: CoverageTools.process_folder: Searching does_not_exist for .jl files...
ERROR: IOError: readdir("does_not_exist"): no such file or directory (ENOENT)
Stacktrace:
 [1] uv_error
   @ ./libuv.jl:97 [inlined]
 [2] readdir(dir::String; join::Bool, sort::Bool)
   @ Base.Filesystem ./file.jl:830
 [3] readdir
   @ ./file.jl:825 [inlined]
 [4] process_folder(folder::String)
   @ CoverageTools ~/.julia/packages/CoverageTools/LiQhZ/src/CoverageTools.jl:261
 [5] top-level scope
   @ REPL[12]:1
```

I just thought the stacktrace might be more obvious this way, if someone is having trouble with the syntax. E.g. if you do `folder1, folder2` instead of `folder1,folder2`, you'll get an error, and if you are puzzled (maybe you don't notice the space right away), by having an explicit upfront error you don't have to look at the source code of CoverageTools to investigate further since you'll get an error first. So that narrows the amount of code to look at (just the few lines in this action). (I think it's generally regarded as good practice to throw errors early for this reason, among others).

**Comment by SaschaMann** [comment]
I've tagged a new release, it should be available now if you're using `v1` (or more specific versions like `v1.2`)

**Comment by ericphanson** [comment]
Thanks!

**Pull request merged**